src/cargo/ops/cargo_new.rs

```diff
@@ -67,22 +64,144 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
                         new_name);
                     try!(config.shell().say(&message, BLACK));
                 }
-                new_name
+                Ok(new_name)
             }
         }
-    };
+    }
+}
+
+fn check_name(name: &str) -> CargoResult<()> {
     for c in name.chars() {
         if c.is_alphanumeric() { continue }
         if c == '_' || c == '-' { continue }
         return Err(human(&format!("Invalid character `{}` in crate name: `{}`",
                                   c, name)));
     }
+    Ok(())
+}
+
+fn detect_source_path_and_type<'a : 'b, 'b>(project_path : &Path, 
+                                            project_name: &str, 
+                                            opts2: &'b mut NewOptions) -> CargoResult<()> {
+    let path = project_path;
+    let name = project_name;
+    
+    let mut found_source_files = 0;
+    
+    if paths::file_already_exists(&path.join("src/main.rs")) {
+        opts2.bin = true;
+        opts2.sourcefile_relative_path = Some(String::from("src/main.rs"));
+        found_source_files += 1;
+    }
+    if paths::file_already_exists(&path.join("main.rs")) {
+        opts2.bin = true;
+        opts2.sourcefile_relative_path = Some(String::from("main.rs"));
+        found_source_files += 1;
+    }
+    fn autodetect_bin_file(p: &Path) -> CargoResult<bool> {
+        let mut content = String::new();
+        try!(File::open(p).and_then(|mut x| x.read_to_string(&mut content)));
+        Ok(content.contains("fn main"))
+    }
+    if paths::file_already_exists(&path.join(format!("{}.rs", name))) {
+        if opts2.bin {
+            // OK
+        } else {
+            opts2.bin = try!(autodetect_bin_file(&path.join(format!("{}.rs", name))));
+            
+        }
+        opts2.sourcefile_relative_path = Some(format!("{}.rs", name));
+        found_source_files += 1;
```

Comment:
This may not be quite the logic we want, this means that if you have `src/main.rs` as well as `src/foo.rs` that you'll get an error about multiple source files, right? The latter may just be the library used by the first though.